0300-Experimental-Xen-hacking-with-GRUB-parameters.patch

```diff
@@ -0,0 +1,366 @@
++#include <grub/xen.h>
++
++#define PARSER_HIT_BACKSLASH 0x1
++#define PARSER_IN_SINGLE_QUOTES 0x2
++#define PARSER_IN_DOUBLE_QUOTES 0x4
++#define PARSER_BASE_WORD_LEN 16
++
++grub_size_t word_list_len;
++char **word_list;
++grub_size_t current_word_len;
++grub_size_t current_word_pos;
++char *current_word;
++char current_char;
++grub_size_t param_dict_len;
++char **param_keys;
++char **param_vals;
++
++static bool
++append_char_to_word (bool allow_null)
++{
++  if (!(current_char >= 0x20) || !(current_char <= 0x7E))
++    {
++      if (current_char != '\0' || !allow_null)
++        {
++          return false;
++        }
++    }
++  if (current_word_pos == current_word_len)
++    {
++      current_word_len *= 2;
++      current_word = grub_realloc (current_word, current_word_len);
++    }
++
++  current_word[current_word_pos] = current_char;
++  current_word_pos++;
++  return true;
++}
++
++static void
++append_word_to_list (void)
++{
++  for (grub_size_t i = 0; i < current_word_pos; i++)
++    {
++      if (current_word[i] <= 0x1F || current_word[i] >= 0x7F)
++        {
++          grub_free (current_word);
++          goto reset_word;
++        }
++    }
++  current_char = '\0';
++  if (!append_char_to_word (true))
++    {
++      grub_error (GRUB_ERR_BUG, N_("couldn't append null terminator to word during Xen cmdline parsing"));
++      grub_print_error ();
++      grub_exit (1);
++    }
++  current_word_len = grub_strlen (current_word) + 1;
++  current_word = grub_realloc (current_word, current_word_len);
++  word_list_len++;
++  word_list = grub_realloc (word_list, word_list_len * sizeof (char *));
++  word_list[word_list_len - 1] = current_word;
++
++reset_word:
++  current_word_len = PARSER_BASE_WORD_LEN;
++  current_word_pos = 0;
++  current_word = grub_malloc (current_word_len);
++}
++
++int
++check_key_is_safe (char *key, grub_size_t len)
++{
++  /*
++   * Ensure only a-z, A-Z, and _ are allowed.
++   */
++  for (grub_size_t i = 0; i < len; i++)
++  {
++    if (! ((key[i] >= 'A' && key[i] <= 'Z')
++          || (key[i] >= 'a' && key[i] <= 'z')
++          || (key[i] == '_') ) )
++      {
++        return 0;
++      }
++  }
++  return 1;
++}
++
++void
++grub_parse_xen_cmdline (void)
++{
++  word_list = grub_malloc (0);
++  current_word_len = PARSER_BASE_WORD_LEN;
++  current_word = grub_malloc (current_word_len);
++  param_keys = grub_malloc (0);
++  param_vals = grub_malloc (0);
++
++  grub_uint8_t parse_flags = 0;
++  char *cmdline = (char *)grub_xen_start_page_addr->cmd_line;
++
++  for (grub_size_t i = 0; i < grub_strlen (cmdline); i++)
++    {
++      current_char = cmdline[i];
++
++      if (parse_flags & PARSER_HIT_BACKSLASH)
++        {
++          parse_flags ^= PARSER_HIT_BACKSLASH;
++          if (!append_char_to_word (false))
++            {
++              goto cleanup;
++            }
++          continue;
++        }
++
++      if (current_char == '\\')
++        {
++          parse_flags ^= PARSER_HIT_BACKSLASH;
```

Comment:
And similar here - use `parse_flags |= PARSER_HIT_BACKSLASH` for setting a bit.

Reply:
Changed.